igvm/drbd.py

```diff
@@ -171,75 +187,72 @@ def replicate_to_slave(self, transaction=None):
-            transaction.on_rollback(
-                'Resume LV device', self.hv.run,
-                'dmsetup resume /dev/{}/{}'
-                .format(self.vg_name, self.lv_name)
+
+            # DRBD is finally up, now replace device which VM talks to on-fly.
+            # In Device Mapper block is always 512 bytes.
+            self.hv.run(
+                'dmsetup load /dev/{}/{} --table "0 {} linear /dev/drbd{} 0"'
+                .format(
+                    self.vg_name, self.lv_name,
+                    dev_size // 512,
+                    self.get_device_minor(),
+                )
             )
-            transaction.on_rollback(
-                'Restore LV device table', self.hv.run,
-                'dmsetup load /dev/{}/{} < {}'
-                .format(self.vg_name, self.lv_name, self.table_file)
+            try:
+                self.hv.run(
+                    'dmsetup resume /dev/{}/{}'
+                    .format(self.vg_name, self.lv_name)
+                )
+            except Exception:
+                # There should be no need for resume because it happens also
+                # via another rollback defined above.  Unfortunately, it is
+                # needed because DRBD won't allow to be shut down when its
+                # device is still held open by somebody.  Also see the comment
+                # about active and inactive slots in stop() method.
+                # WARNING: Potential race between writes to DRBD and underlying
+                # device - potential data loss?
+                # TODO: suspend VM for rollback
+                self.hv.run(
+                    'dmsetup load /dev/{}/{} < {}'
+                    .format(self.vg_name, self.lv_name, self.table_file)
+                )
+                self.hv.run(
+                    'dmsetup resume /dev/{}/{}'
+                    .format(self.vg_name, self.lv_name)
+                )
+                raise
+        except Exception:
+            # The "up" command might fail due to misconfiguration but the
+            # device is started nevertheless. This is why "down" rollback is
+            # always performed.
+            self.hv.run('drbdadm down {}'.format(self.vm_name))
+            self.hv.run(
+                'dmsetup resume /dev/{}/{}'.format(self.vg_name, self.lv_name)
             )
-
-        self.hv.run('dmsetup resume /dev/{}/{}'.format(
-            self.vg_name, self.lv_name
-        ))
+            raise
 
     def replicate_from_master(self, transaction=None):
         self.hv.run('drbdadm create-md {}'.format(self.vm_name))
         self.hv.run('drbdadm up {}'.format(self.vm_name))
-        transaction.on_rollback(
-            'Bring DRBD device down', self.hv.run,
-            'drbdadm down {}'.format(self.vm_name)
-        )
-        self.hv.run('drbdadm wait-connect {}'.format(self.vm_name))
+        try:
+            self.hv.run('drbdadm wait-connect {}'.format(self.vm_name))
+        except Exception:
```

Comment:
I'm pretty sure run would raise a SystemExit on failure which is not a subclass of Exception but subclasses BaseException directly. This would therefore not work they way you expect, I think. See: https://github.com/fabric/fabric/issues/277#issuecomment-1861413

Reply:
Have we addressed this at all in igvm so far?